sidecar/sidecar.go

```diff
@@ -260,56 +260,45 @@ func buildServiceRules(conf *config.Config, kubeClient kubernetes.Interface) (ap
 	}
 }
 
-func startProxy(conf *config.Config, discovery api.ServiceDiscovery, kubeClient kubernetes.Interface) error {
-	var err error
-
-	service := nginx.NewService(conf.Service.Name, conf.Service.Tags)
-	if err := service.Start(); err != nil {
-		logrus.WithError(err).Error("NGINX service failed to start")
-		return err
+func buildProxyAdapter(conf *config.Config, discovery monitor.DiscoveryMonitor, rules monitor.RulesMonitor) (proxy.Adapter, error) {
+	switch conf.ProxyAdapter {
+	case config.NGINXAdapter:
+		return proxy.NewNGINXAdapter(conf, discovery, rules), nil
+	default:
+		return nil, fmt.Errorf("unsupported proxy adapter: %v", conf.ProxyAdapter)
 	}
+}
 
-	nginxClient := nginx.NewClient("http://localhost:5813")
-	nginxManager := nginx.NewManager(
-		nginx.Config{
-			Client: nginxClient,
-		},
-	)
-	nginxProxy := proxy.NewNGINXProxy(nginxManager)
-
+func startProxy(conf *config.Config, discovery api.ServiceDiscovery, kubeClient kubernetes.Interface) error {
 	rules, err := buildServiceRules(conf, kubeClient)
 	if err != nil {
 		logrus.WithError(err).Error("Could not create service rules client")
 		return err
 	}
 
 	rulesMonitor := monitor.NewRulesMonitor(monitor.RulesConfig{
-		Rules: rules,
-		Listeners: []monitor.RulesListener{
-			nginxProxy,
-		},
+		Rules:        rules,
 		PollInterval: conf.A8Controller.Poll,
 	})
 
 	discoveryMonitor := monitor.NewDiscoveryMonitor(monitor.DiscoveryConfig{
 		Discovery: discovery,
-		Listeners: []monitor.DiscoveryListener{
-			nginxProxy,
-		},
 	})
 
-	go func() {
-		if err = rulesMonitor.Start(); err != nil {
-			logrus.WithError(err).Error("Controller monitor failed")
-		}
-	}()
-	go func() {
-		if err = discoveryMonitor.Start(); err != nil {
-			logrus.WithError(err).Error("Registry monitor failed")
-		}
-	}()
+	proxyAdapter, err := buildProxyAdapter(conf, discoveryMonitor, rulesMonitor)
+	if err != nil {
+		logrus.WithError(err).Error("Could not build proxy adapter")
+		return err
+	}
+
+	if err := proxyAdapter.Start(); err != nil {
+		logrus.WithError(err).Error("Could not start proxy adapter")
+		return err
+	}
 
-	debugger := debug.NewAPI(nginxProxy)
+	debugger := debug.NewAPI()
+	rulesMonitor.AddListener(debugger)
+	discoveryMonitor.AddListener(debugger)
 
 	a := rest.NewApi()
```

Comment:
Unrelated to this PR, but the configuration of the debug API server (lines 303-319) should probably go into the `debug` package (e.g., as part of `debug.NewAPI()`).